openstack/02a-MAClow/IEEE802154_security.c

```diff
@@ -0,0 +1,887 @@
+   security_vars.MacKeyTable.KeyDescriptorElement[0].DeviceTable = &security_vars.MacDeviceTable;
+}
+
+//=========================== public ==========================================
+/**
+\brief Adding of Auxiliary Security Header to the IEEE802.15.4 MAC header
+*/
+void IEEE802154security_prependAuxiliarySecurityHeader(OpenQueueEntry_t* msg){
+
+   bool frameCounterSuppression;
+   uint8_t temp8b;
+   open_addr_t* temp_keySource;
+   uint8_t auxiliaryLength;
+
+   frameCounterSuppression = 1; //the frame counter is carried in the frame
+
+   //max length of MAC frames
+   // length of authentication Tag
+   msg->l2_authenticationLength = IEEE802154security_authLengthChecking(msg->l2_securityLevel);
+
+   //length of auxiliary security header
+   auxiliaryLength = IEEE802154security_auxLengthChecking(msg->l2_keyIdMode,
+                                                          frameCounterSuppression,
+                                                          security_vars.m_macFrameCounterMode); //length of Key ID field
+
+
+   if ((msg->length+auxiliaryLength+msg->l2_authenticationLength+2) >= 130 ){ //2 bytes of CRC, 130 MaxPHYPacketSize
+      return;
+   }
+
+   //start setting the Auxiliary Security Header
+   if (msg->l2_keyIdMode !=0){//if the KeyIdMode is zero, keyIndex and KeySource are omitted
+      temp8b = msg->l2_keyIndex; //key index field
+      packetfunctions_reserveHeaderSize(msg, sizeof(uint8_t));
+      *((uint8_t*)(msg->payload)) = temp8b;
+   }
+
+   switch (msg->l2_keyIdMode){
+      case 0: //no KeyIDMode field - implicit
+    	 temp_keySource = &security_vars.m_macDefaultKeySource;
+         memcpy(&(msg->l2_keySource), temp_keySource, sizeof(open_addr_t));
+         break;
+      case 1:// macDefaultKeySource
+         msg->l2_keySource = security_vars.m_macDefaultKeySource;
+         break;
+      case 2: //keySource with 16b address
+         temp_keySource = &msg->l2_keySource;
+         packetfunctions_reserveHeaderSize(msg, sizeof(uint8_t));
+         *((uint8_t*)(msg->payload)) = temp_keySource->addr_64b[6];
+         packetfunctions_reserveHeaderSize(msg, sizeof(uint8_t));
+         *((uint8_t*)(msg->payload)) = temp_keySource->addr_64b[7];
+         break;
+      case 3: //keySource with 64b address
+         temp_keySource = &msg->l2_keySource;
+         packetfunctions_writeAddress(msg,temp_keySource,OW_LITTLE_ENDIAN);
+         break;
+      default:
+         return;
+   }
+
+   //Frame Counter
+   if (frameCounterSuppression==0){
+      //here I have to insert the ASN: I can only reserve the space and
+      //save the pointer. The ASN will be added by IEEE802.15.4e
+
+      // reserve space
+      packetfunctions_reserveHeaderSize(msg,sizeof(macFrameCounter_t));
+
+      // Keep a pointer to where the ASN will be
+      // Note: the actual value of the current ASN will be written by the
+      //    IEEE802.15.4e when transmitting
+      msg->l2_ASNpayload = msg->payload;
```

Comment:
@changtengfei thank you for your finding! 

I will update the code as soon as possible.